discovery_utils/utils/viz_landscape.py

```diff
@@ -0,0 +1,546 @@
+    )
+
+    # Fit the model using precomputed embeddings
+    topics, probs = topic_model.fit_transform(
+        vectors_df["text"],
+        embeddings=np.array(vectors_df["vector"].to_list()),
+    )
+    vectors_df["topics"] = topics
+
+    # Assign noise points to other topics
+    try:
+        reduced_topics = topic_model.reduce_outliers(
+            vectors_df["text"].to_list(),
+            topics=vectors_df["topics"].to_list(),
+            strategy="embeddings",
+            embeddings=np.array(vectors_df["vector"].to_list()),
+        )
+        vectors_df["reduced_topics"] = reduced_topics
+        logging.info("Outliers were successfully reduced")
+    except Exception as e:
+        vectors_df["reduced_topics"] = topics
+        logging.error(f"An error occurred while reducing outliers: {e}")
+    vectors_df = vectors_df.assign(reduced_topics=lambda df: df.reduced_topics.astype(int))
+    return vectors_df, topic_model
+
+
+def generate_landscape_keywords(
+    viz_df: pd.DataFrame,
+    n_keyword_clusters: int = N_KEYWORD_CLUSTERS,
+    random_state: int = 10,
+) -> Tuple[pd.DataFrame, pd.DataFrame]:
+    """Generate keywords for the landscape"""
+    clusterer = KMeans(n_clusters=n_keyword_clusters, random_state=random_state)
+    clusterer.fit(viz_df[["umap_x", "umap_y"]])
+    soft_clusters = list(clusterer.labels_)
+
+    title_texts = viz_df["text"].apply(preproc)
+    _cluster_texts = cluster_texts(title_texts, soft_clusters)
+    _cluster_keywords = cluster_keywords(
+        documents=list(_cluster_texts.values()),
+        cluster_labels=list(_cluster_texts.keys()),
+        n=2,
+        max_df=0.90,
+        min_df=0.01,
+        Vectorizer=TfidfVectorizer,
+    )
+    viz_df["soft_cluster"] = soft_clusters
+    viz_df["soft_cluster_"] = [str(x) for x in soft_clusters]
+    viz_df["keyword_cluster"] = viz_df["soft_cluster"].apply(lambda x: ", ".join(_cluster_keywords[x]))
+
+    centroids = (
+        viz_df.groupby("soft_cluster")
+        .agg(x_c=("umap_x", "mean"), y_c=("umap_y", "mean"))
+        .reset_index()
+        .assign(keywords=lambda x: x.soft_cluster.apply(lambda y: ", ".join(_cluster_keywords[y])))
+    )
+
+    return viz_df.drop(["soft_cluster", "soft_cluster_"], axis=1), centroids
+
+
+def chart_keywords(centroids: pd.DataFrame) -> alt.Chart:
+    """Generate a chart with keywords"""
+    return (
+        alt.Chart(centroids)
+        .mark_text(
+            fontSize=13.5,
+            fontStyle="bold",
+            opacity=0.8,
+            stroke="white",
+            strokeWidth=1,
+            strokeOffset=0,
+            strokeOpacity=0.4,
+        )
+        .encode(x=alt.X("x_c:Q"), y=alt.Y("y_c:Q"), text=alt.Text("keywords"))
+    )
+
+
+def generate_reduced_embeddings(vectors_df: pd.DataFrame) -> np.ndarray:
+    """Generate reduced embeddings for visualisation purposes"""
+    return UMAP(n_neighbors=15, n_components=2, min_dist=0.0, metric="cosine").fit_transform(
```

Comment:
You could add a random seed to UMAP too, just to make the outputs as reproducible as possible
```suggestion
    return UMAP(n_neighbors=15, n_components=2, min_dist=0.0, metric="cosine", random_state=42).fit_transform(
```

Reply:
Good idea!